tooling/noir_js/test/backend/barretenberg.ts

```diff
@@ -0,0 +1,122 @@
+
+    this.api = api;
+    this.acirComposer = composer;
+  }
+
+  async initBarretenberg(numThreads: number, acirUncompressedBytecode: Uint8Array) {
+    const api = await Barretenberg.new(numThreads);
+
+    const [exact, total, subgroupSize] = await api.acirGetCircuitSizes(acirUncompressedBytecode);
+    const crs = await Crs.new(subgroupSize + 1);
+    await api.commonInitSlabAllocator(subgroupSize);
+    await api.srsInitSrs(new RawBuffer(crs.getG1Data()), crs.numPoints, new RawBuffer(crs.getG2Data()));
+
+    const acirComposer = await api.acirNewAcirComposer(subgroupSize);
+    return { api: api, composer: acirComposer };
+  }
+
+  // Generate an outer proof. This is the proof for the circuit which will verify
+  // inner proofs and or can be seen as the proof created for regular circuits.
+  //
+  // The settings for this proof are the same as the settings for a "normal" proof
+  // ie one that is not in the recursive setting.
+  async generateOuterProof(decompressedWitness: Uint8Array) {
+    const makeEasyToVerifyInCircuit = false;
+    return this.generateProof(decompressedWitness, makeEasyToVerifyInCircuit);
+  }
+
+  // Generates an inner proof. This is the proof that will be verified
+  // in another circuit.
+  //
+  // This is sometimes referred to as a recursive proof.
+  // We avoid this terminology as the only property of this proof
+  // that matters, is the fact that it is easy to verify in another
+  // circuit. We _could_ choose to verify this proof in the CLI.
+  //
+  // We set `makeEasyToVerifyInCircuit` to true, which will tell the backend to
+  // generate the proof using components that will make the proof
+  // easier to verify in a circuit.
+  async generateInnerProof(witness: Uint8Array) {
+    const makeEasyToVerifyInCircuit = true;
+    return this.generateProof(witness, makeEasyToVerifyInCircuit);
+  }
+
+  async generateProof(decompressedWitness: Uint8Array, makeEasyToVerifyInCircuit: boolean) {
+    const proof = await this.api.acirCreateProof(
+      this.acirComposer,
+      this.acirUncompressedBytecode,
+      decompressedWitness,
+      makeEasyToVerifyInCircuit,
+    );
+
+    return proof;
+  }
+
+  // Generates artifacts that will be passed to a circuit that will verify this proof.
+  //
+  // Instead of passing the proof and verification key as a byte array, we pass them
+  // as fields which makes it cheaper to verify in a circuit.
+  //
+  // The proof that is passed here will have been created using the `generateInnerProof`
+  // method.
+  //
+  // The number of public inputs denotes how many public inputs are in the inner proof.
+  async generateInnerProofArtifacts(proof: Uint8Array, numOfPublicInputs: number = 0) {
+    const proofAsFields = await this.api.acirSerializeProofIntoFields(this.acirComposer, proof, numOfPublicInputs);
+
+    // TODO: perhaps we should put this in the init function. Need to benchmark
+    // TODO how long it takes.
```

Comment:
TODO: if we do decide to allow this class to change its circuit, then we'd want it to be here instead of in init which doesn't get called again.

Reply:
It would be faster to read in an already generated verification key as it is small, but generating a verification key requires generation of the proving key.